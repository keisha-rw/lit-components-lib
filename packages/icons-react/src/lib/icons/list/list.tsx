import * as React from 'react';
import { PdsIconList as PdsIconListElement } from '@keisha/design-system-icons-web';
import {
  IconSize,
  ValidIconColor,
} from '../../../utils/icon-utilities';
import { createPdsReactComponent } from '../../../../../react/src/lib/create-pds-react-component';

interface IconProps {
  size?: IconSize;
  color?: ValidIconColor;
}

export const PdsIconList = createPdsReactComponent<IconProps>()({
  tagName: 'pds-icon-list',
  elementClass: PdsIconListElement,
  react: React,
});
