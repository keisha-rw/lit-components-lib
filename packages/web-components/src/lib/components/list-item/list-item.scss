/*------------------------------------*\
 #LIST_ITEM
\*------------------------------------*/

.pds-c-list-item {
  display: flex;
}

.pds-c-list-item__icon-wrapper {
  margin-right: size(1);
  margin-top: size(0.75);
}

.pds-c-list-item__term {
  @include pds-semantic-label-lg;
}

.pds-c-list-item__details {
  @include pds-semantic-body-default-compact;

  margin: 0;
}

.pds-c-list-item__term-reverse {
  @include pds-semantic-body-default-compact;
}

.pds-c-list-item__details-reverse {
  @include pds-semantic-label-lg;

  margin: 0;
}
