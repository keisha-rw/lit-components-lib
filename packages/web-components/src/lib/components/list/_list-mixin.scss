/*------------------------------------*\
 #LIST_MIXIN
\*------------------------------------*/

@mixin list {
  list-style: none;
  padding: 0;
  margin: 0;
  display: flex;
  flex-flow: column wrap;
  gap: size(3);
}
