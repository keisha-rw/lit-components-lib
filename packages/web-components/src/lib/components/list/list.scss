@import './list-mixin';

/*------------------------------------*\
 #LIST
\*------------------------------------*/

.pds-c-list {
  @include pds-semantic-body-lg;
  @include list;
}

::slotted(pds-list-item) {
  // Removes stretch from list items within to use their natural width
  flex: 0;
}

.pds-c-list--sm {
  @include pds-semantic-body-default-compact;
}

:host {
  display: block;
}

.pds-c-list--horizontal {
  flex-flow: row wrap;
}

.pds-c-list--align-center {
  align-items: center;
}

.pds-c-list--horizontal.pds-c-list--align-center {
  justify-content: center;
}

.pds-c-list--spacing-sm {
  gap: size(1);
}

.pds-c-list--spacing-none {
  gap: 0;
}

.pds-c-list--spacing-lg {
  gap: size(5);
}
