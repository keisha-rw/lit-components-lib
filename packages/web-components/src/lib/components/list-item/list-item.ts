import { html } from 'lit';
import { property, queryAssignedElements } from 'lit/decorators.js';
import { pdsCustomElement as customElement } from '../../decorators/pds-custom-element';
import { PdsElement } from '../PdsElement';
// eslint-disable-next-line import/no-cycle
import { size, ListSize } from '../list/list';
import styles from './list-item.scss?inline';

/**
 * @summary A list item (li) element
 *
 * @slot default The content of the list item
 * @slot icon This slot holds an icon
 */
@customElement('pds-list-item', {
  category: 'component',
  type: 'component',
  state: 'stable',
  styles,
})
export class PdsListItem extends PdsElement {
  /**
   * Size
   * Should be inherited from pds-list
   * If provided directly, renders the list with small typography
   * if icons are in the list item, they will inherit the size.
   */
  @property({ reflect: true })
  size: ListSize = 'md';

  @property()
  variant: 'default' | 'description' | 'description-reverse' = 'default';

  /**
   * This grabs the icon from the slot
   * @internal
   */
  @queryAssignedElements({ slot: 'icon' })
  listItemIcons: HTMLElement[];

  addSizeTolistItemIcons() {
    if (
      this.listItemIcons &&
      this.listItemIcons.length &&
      size.includes(this.size)
    ) {
      this.listItemIcons.forEach((icon) =>
        icon.setAttribute('size', this.size),
      );
    }
  }

  updated() {
    this.addSizeTolistItemIcons();
  }

  render() {
    if (
      this.variant === 'description' ||
      this.variant === 'description-reverse'
    ) {
      return html`<dt
          class=${this.variant === 'description-reverse'
            ? this.classEl('term-reverse')
            : this.classEl('term')}
        >
          <slot name="description-term"></slot>
        </dt>
        <dd
          class="${this.variant === 'description-reverse'
            ? this.classEl('details-reverse')
            : this.classEl('details')}"
        >
          <slot name="description"></slot>
        </dd>`;
    }
    return html`<li class=${this.getClass()} role="listitem">
      ${this.slotNotEmpty('icon')
        ? html`<span class="${this.classEl('icon-wrapper')}"
            ><slot name="icon" @slotchange=${this.addSizeTolistItemIcons}></slot
          ></span>`
        : ''}
      <slot></slot>
    </li>`;
  }
}
