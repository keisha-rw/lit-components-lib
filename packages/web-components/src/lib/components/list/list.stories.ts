import { html, nothing } from 'lit';
import { Meta, StoryObj } from '@storybook/web-components';
import './list';
import '../link/link';
import '../../../../.storybook/components/placeholder-element/placeholder-element';
import '@keisha/design-system-icons-web/check';

export default {
  title: 'Components/List',
  component: 'pds-list',
  tags: ['autodocs'],
  argTypes: {
    size: {
      control: 'radio',
      options: ['md', 'sm'],
    },
    spacing: {
      control: 'radio',
      options: ['sm', 'md', 'lg', 'none'],
    },
    align: {
      control: 'radio',
      options: ['center', 'default'],
    },
    orientation: {
      control: 'radio',
      options: ['horizontal', 'default'],
    },
    variant: {
      control: 'select',
      options: ['default', 'description', 'description-reverse'],
    },
  },
  render: (args) => {
    if (args.variant === 'description') {
      return html`<pds-list variant=${args.variant || nothing}>
        <pds-list-item
          ><span slot="description">List item description</span
          ><span slot="description-term">List item label</span></pds-list-item
        >
        <pds-list-item>
          <span slot="description">List item description</span
          ><span slot="description-term">List item label</span></pds-list-item
        >
        <pds-list-item>
          <span slot="description">List item description</span
          ><span slot="description-term">List item label</span></pds-list-item
        >
        <pds-list-item>
          <span slot="description">List item description</span
          ><span slot="description-term">List item label</span></pds-list-item
        >
        <pds-list-item>
          <span slot="description">List item description</span
          ><span slot="description-term">List item label</span></pds-list-item
        >
      </pds-list>`;
    }
    if (args.variant === 'description-reverse') {
      return html`<pds-list variant=${args.variant || nothing}>
        <pds-list-item
          ><span slot="description">List item description</span
          ><span slot="description-term">List item label</span></pds-list-item
        >
        <pds-list-item>
          <span slot="description">List item description</span
          ><span slot="description-term">List item label</span></pds-list-item
        >
        <pds-list-item>
          <span slot="description">List item description</span
          ><span slot="description-term">List item label</span></pds-list-item
        >
        <pds-list-item>
          <span slot="description">List item description</span
          ><span slot="description-term">List item label</span></pds-list-item
        >
        <pds-list-item>
          <span slot="description">List item description</span
          ><span slot="description-term">List item label</span></pds-list-item
        >
      </pds-list>`;
    }
    if (args.variant === 'default') {
      return html`<pds-list
        size=${args.size || nothing}
        align=${args.align || nothing}
        orientation=${args.orientation || nothing}
        spacing=${args.spacing || nothing}
        variant=${args.variant || nothing}
      >
        <pds-list-item><span>List item label</span></pds-list-item>
        <pds-list-item><span>List item label</span></pds-list-item>
        <pds-list-item><span>List item label</span></pds-list-item>
      </pds-list>`;
    }
    return html`<pds-list
      size=${args.size || nothing}
      align=${args.align || nothing}
      orientation=${args.orientation || nothing}
      spacing=${args.spacing || nothing}
      variant=${args.variant || nothing}
    >
      ${args.child || nothing}
    </pds-list>`;
  },
} as Meta;

// span element is needed to wrap plain text slots https://github.com/storybookjs/storybook/issues/15963

export const Default: StoryObj = {
  name: 'Default',
  args: {
    variant: 'default',
  },
  parameters: {
    a11y: {
      config: {
        rules: [
          {
            // Color contrast is sufficient - this is a false positive.
            //
            // @link https://dequeuniversity.com/rules/axe/4.6/color-contrast
            id: 'color-contrast',
            enabled: false,
          },
        ],
      },
    },
  },
};

export const DefaultSmall: StoryObj = {
  name: 'Default small',
  args: {
    size: 'sm',
    child: html`<pds-list-item><span>List item label</span></pds-list-item>
      <pds-list-item><span>List item label</span></pds-list-item>
      <pds-list-item><span>List item label</span></pds-list-item>`,
  },
};

export const DefaultWithIcons: StoryObj = {
  name: 'Default with icons',
  args: {
    child: html`<pds-list-item>
      <pds-icon-check slot="icon"></pds-icon-check>
      <span
        >This is a really really long list item with a lot of text so it can
        showcase how the text wraps with the icon on mobile. </span
      ></pds-list-item>
      <pds-list-item>
        <pds-icon-check slot="icon"></pds-icon-check>
        <span>This is a list item.</span></pds-list-item
      ><pds-list-item>
          <pds-icon-check slot="icon"></pds-icon-check>
        <span>This is a list item.</span>
      </pds-list-item></pds-list-item
    >`,
  },
};

export const DefaultSmallSpacing: StoryObj = {
  name: 'Default small spacing',
  args: {
    spacing: 'sm',
    child: html`<pds-list-item
        ><placeholder-element
          ><span>List item label</span></placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          ><span>List item label</span></placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          ><span>List item label</span></placeholder-element
        ></pds-list-item
      >`,
  },
};

export const Horizontal: StoryObj = {
  name: 'Horizontal',
  args: {
    orientation: 'horizontal',
    child: html`<pds-list-item>List item label</pds-list-item>
      <pds-list-item>List item label</pds-list-item>
      <pds-list-item>List item label</pds-list-item>
      <pds-list-item>List item label</pds-list-item>
      <pds-list-item>List item label</pds-list-item>
      <pds-list-item>List item label</pds-list-item>
      <pds-list-item>List item label</pds-list-item>
      <pds-list-item>List item label</pds-list-item>
      <pds-list-item>List item label</pds-list-item>
      <pds-list-item>List item label</pds-list-item>`,
  },
};

export const HorizontalSmall: StoryObj = {
  name: 'Horizontal small',
  args: {
    orientation: 'horizontal',
    size: 'sm',
    child: html`<pds-list-item>List item label</pds-list-item>
      <pds-list-item>List item label</pds-list-item>
      <pds-list-item>List item label</pds-list-item>
      <pds-list-item>List item label</pds-list-item>
      <pds-list-item>List item label</pds-list-item>
      <pds-list-item>List item label</pds-list-item>
      <pds-list-item>List item label</pds-list-item>
      <pds-list-item>List item label</pds-list-item>
      <pds-list-item>List item label</pds-list-item>
      <pds-list-item>List item label</pds-list-item>`,
  },
};

export const HorizontalSmallSpacing: StoryObj = {
  name: 'Horizontal small spacing',
  args: {
    orientation: 'horizontal',
    spacing: 'sm',
    child: html`<pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >`,
  },
};

export const HorizontalLargeSpacing: StoryObj = {
  name: 'Horizontal large spacing',
  args: {
    orientation: 'horizontal',
    spacing: 'lg',
    child: html`<pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >`,
  },
};

export const AlignCenter: StoryObj = {
  name: 'Align center',
  args: {
    align: 'center',
    child: html`<pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >`,
  },
};

export const AlignCenterHorizontal: StoryObj = {
  name: 'Align center horizontal',
  args: {
    orientation: 'horizontal',
    align: 'center',
    child: html`<pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >
      <pds-list-item
        ><placeholder-element
          >List item label</placeholder-element
        ></pds-list-item
      >`,
  },
};

export const LinkList: StoryObj = {
  name: 'Link list',
  args: {
    child: html`<pds-list-item
        ><pds-link href="#">This is a link</pds-link></pds-list-item
      >
      <pds-list-item
        ><pds-link href="#">This is a link</pds-link></pds-list-item
      >
      <pds-list-item
        ><pds-link href="#">This is a link</pds-link></pds-list-item
      >
      <pds-list-item
        ><pds-link href="#">This is a link</pds-link></pds-list-item
      >
      <pds-list-item
        ><pds-link href="#">This is a link</pds-link></pds-list-item
      >
      <pds-list-item
        ><pds-link href="#">This is a link</pds-link></pds-list-item
      >`,
  },
};

export const DescriptionDefault: StoryObj = {
  name: 'Description default',
  args: { variant: 'description' },
  parameters: {
    a11y: {
      config: {
        rules: [
          {
            // Description list items are li elements, so this check is invalid.
            //
            // @link https://dequeuniversity.com/rules/axe/4.4/definition-list
            id: 'definition-list',
            enabled: false,
          },
          {
            // The shadow dom makes this check difficult, but the dd and dt items are in a dl.
            //
            // @link https://dequeuniversity.com/rules/axe/4.4/dlitem
            id: 'dlitem',
            enabled: false,
          },
        ],
      },
    },
  },
};

export const DescriptionReverse: StoryObj = {
  name: 'Description reverse',
  args: {
    variant: 'description-reverse',
  },
  parameters: {
    a11y: {
      config: {
        rules: [
          {
            // Description list items are li elements, so this check is invalid.
            //
            // @link https://dequeuniversity.com/rules/axe/4.4/definition-list
            id: 'definition-list',
            enabled: false,
          },
          {
            // The shadow dom makes this check difficult, but the dd and dt items are in a dl.
            //
            // @link https://dequeuniversity.com/rules/axe/4.4/dlitem
            id: 'dlitem',
            enabled: false,
          },
        ],
      },
    },
  },
};
