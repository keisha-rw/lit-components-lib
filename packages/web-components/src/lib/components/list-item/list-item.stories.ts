import { html, nothing } from 'lit';
import { Meta, StoryObj } from '@storybook/web-components';
import '../list/list';
import './list-item';
import '@keisha/design-system-icons-web/check';

export default {
  title: 'Components/List/List item',
  component: 'pds-list-item',
  tags: ['autodocs'],
  argTypes: {
    variant: {
      control: 'select',
      options: ['default', 'description', 'description-reverse'],
    },
  },
  parameters: {
    happo: false,
    docs: {
      description: {
        component:
          'Note: This component is ONLY meant to be used as a subcomponent of the list component.',
      },
    },
    a11y: {
      config: {
        rules: [
          {
            // This is not an issue when it's contained properly within the pds-list element
            //
            // @link https://dequeuniversity.com/rules/axe/4.4/aria-required-parent
            id: 'aria-required-parent',
            enabled: false,
          },
          {
            // The shadow dom makes this check difficult, but the dd and dt items are in a dl.
            //
            // @link https://dequeuniversity.com/rules/axe/4.4/dlitem
            id: 'dlitem',
            enabled: false,
          },
        ],
      },
    },
  },
  render: (args) => {
    if (
      args.variant === 'description' ||
      args.variant === 'description-reverse'
    ) {
      return html`<pds-list-item variant="${args.variant || nothing}"
        ><span slot="description">List item description</span
        ><span slot="description-term">List item label</span></pds-list-item
      >`;
    }
    return html`<pds-list-item variant="${args.variant || nothing}"
      >This is a list item.</pds-list-item
    > `;
  },
} as Meta;

// span element is needed to wrap plain text slots https://github.com/storybookjs/storybook/issues/15963

export const ListItem: StoryObj = {
  name: 'Default',
  args: {
    text: 'List item',
    args: { variant: 'default' },
  },
};

export const ListItemWithIcon: StoryObj = {
  name: 'Icon',
  args: {
    text: 'List item with icon',
  },
  render: () =>
    html`<pds-list-item>
      <pds-icon-check slot="icon"></pds-icon-check>
      <slot>This is a list item.</slot>
    </pds-list-item>`,
};

export const Default: StoryObj = {
  name: 'Description Default',
  args: { variant: 'description' },
};

export const Reverse: StoryObj = {
  name: 'Description Reverse',
  args: {
    variant: 'description-reverse',
  },
};
