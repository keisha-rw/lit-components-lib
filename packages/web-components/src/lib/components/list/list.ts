/* eslint-disable import/no-duplicates */
import { html } from 'lit';
import { property, queryAssignedElements } from 'lit/decorators.js';
import { pdsCustomElement as customElement } from '../../decorators/pds-custom-element';
import { PdsElement } from '../PdsElement';
import styles from './list.scss?inline';
// eslint-disable-next-line import/no-cycle
import { PdsListItem } from '../list-item/list-item';
import '../list-item/list-item';

export const size = ['sm', 'md'] as const;
export type ListSize = (typeof size)[number];

/**
 * @summary An unordered list element (ul)
 *
 * @slot default Contains one or more list items within the pds-list, restricted to pds-list-item elements
 */
@customElement('pds-list', {
  category: 'component',
  type: 'component',
  state: 'stable',
  styles,
})
export class PdsList extends PdsElement {
  /**
   * Size property
   * sm - Renders the list and child list items with small typography
   * also adjusts the icon size if provided
   */
  @property()
  size: ListSize = 'md';

  /**
   * variant
   * - **default** - renders the list with default typography styles
   * - **description** - renders the list with description typography styles
   * - **description-reverse** - renders the list with reverse typography styles
   */
  @property()
  variant: 'default' | 'description' | 'description-reverse' = 'default';

  @property()
  spacing: 'sm' | 'md' | 'lg' | 'none' = 'md';

  /**
   * Align variant
   * - Default section is left aligned
   * - **center** horizontally center aligns the list
   */
  @property()
  align: 'center' | 'default' = 'default';

  /**
   * Style variant
   * - **primary** renders the list used for primary actions
   */
  @property()
  orientation: 'horizontal' | 'default' = 'default';

  /**
   * @internal
   */
  get classNames() {
    return {
      /* This is equivalent to doing
       * 'primary': this.variant === 'primary',
       * 'secondary': this.variant === 'secondary',
       */
      [`${this.orientation}`]: !!this.orientation,
      [`align-${this.align}`]: !!this.align,
      [`${this.size}`]: !!this.size,
      [`spacing-${this.spacing}`]: !!this.spacing,
      [this.variant]: !!this.variant,
    };
  }

  /**
   * This grabs the listItem slots
   * @internal
   */
  @queryAssignedElements({ slot: undefined, selector: 'pds-list-item' })
  listItems: HTMLElement[];

  addSizeToListItems() {
    if (this.listItems && this.listItems.length && size.includes(this.size)) {
      this.listItems.forEach((listItem) =>
        listItem.setAttribute('size', this.size),
      );
    }
  }

  handleSlotChange(e: Event) {
    this.handleSlotValidation(e);
    this.addSizeToListItems();
  }

  updated() {
    this.addSizeToListItems();
  }

  setChildVariant() {
    const descriptionListItem = Array.from(
      this.querySelectorAll('pds-list-item'),
    ) as PdsListItem[];
    descriptionListItem.forEach((item) => {
      // eslint-disable-next-line no-param-reassign
      item.variant = this.variant;
    });
  }

  render() {
    if (
      this.variant === 'description' ||
      this.variant === 'description-reverse'
    ) {
      return html`<dl class=${this.getClass()}>
        <slot @slotchange=${this.setChildVariant}></slot>
      </dl>`;
    }
    return html`<ul class=${this.getClass()} part="list" role="list">
      <slot
        allowed-elements="pds-list-item"
        @slotchange=${this.handleSlotChange}
      ></slot>
    </ul>`;
  }
}
