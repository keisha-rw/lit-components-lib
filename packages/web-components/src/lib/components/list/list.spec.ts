import * as path from 'path';
import initStoryshots from '@storybook/addon-storyshots';
import { assert, fixture } from '@open-wc/testing';
import { PdsList } from './list';
import createSnapshots from '../../../../test/utils/snapshots';
import { PdsListItem } from '../list-item/list-item';

initStoryshots({
  storyKindRegex: /List/,
  framework: 'web-components',
  configPath: path.join(__dirname, '../../../../.storybook'),
  asyncJest: true,
  test: createSnapshots,
});

describe('List unit tests', () => {
  let element: Element;

  beforeEach(async () => {
    element = await fixture('<pds-list/>');
  });

  it('is an instance of PdsList', () => {
    expect(element).toBeInstanceOf(PdsList);
  });

  it('Passes size down to list item', async () => {
    element = await fixture('<pds-list><pds-list-item /></pds-list>');

    const listItemSlot = element.querySelector('pds-list-item') as PdsListItem;
    expect(listItemSlot.getAttribute('size')).toEqual('md');

    element = await fixture('<pds-list size="sm"><pds-list-item /></pds-list>');

    const smListItemSlot = element.querySelector(
      'pds-list-item',
    ) as PdsListItem;
    expect(smListItemSlot.getAttribute('size')).toEqual('sm');
  });

  it('Defaults to md when a bad size is passed', async () => {
    element = await fixture('<pds-list size="sm"><pds-list-item /></pds-list>');

    const listItemSlot = element.querySelector('pds-list-item') as PdsListItem;
    expect(listItemSlot.getAttribute('size')).toEqual('sm');

    element = await fixture('<pds-list size="xl"><pds-list-item /></pds-list>');

    const smListItemSlot = element.querySelector(
      'pds-list-item',
    ) as PdsListItem;
    expect(smListItemSlot.getAttribute('size')).toEqual('md');
  });

  it('should set child variant to "default"', async () => {
    element = await fixture(
      '<pds-list variant="default"><pds-list-item /></pds-list>',
    );
    const listItems = Array.from(
      element.querySelectorAll('pds-list-item'),
    ) as PdsListItem[];
    listItems.forEach((item) => {
      expect(item.variant).toEqual('default');
    });
  });

  it('should set child variant to "description"', async () => {
    element = await fixture(
      '<pds-list variant="description"><pds-list-item /></pds-list>',
    );
    const listItems = Array.from(
      element.querySelectorAll('pds-list-item'),
    ) as PdsListItem[];
    listItems.forEach((item) => {
      expect(item.variant).toEqual('description');
    });
  });

  it('should set child variant to "description-reverse"', async () => {
    element = await fixture(
      '<pds-list variant="description-reverse"><pds-list-item /></pds-list>',
    );
    const listItems = Array.from(
      element.querySelectorAll('pds-list-item'),
    ) as PdsListItem[];
    listItems.forEach((item) => {
      expect(item.variant).toEqual('description-reverse');
    });
  });

  it('is accessible', async () => {
    await assert.isAccessible(element, {
      ignoredRules: ['aria-required-parent'],
    });
  });

  it('populates default slot correctly with type pds-list-item', async () => {
    const consoleError = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});
    element = await fixture('<pds-list size="sm"><pds-list-item/><pds-list>');
    expect(consoleError).not.toHaveBeenCalled();
    consoleError.mockRestore();
  });

  it('does not populate the default slot with an element other than pds-list-item', async () => {
    const consoleError = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});
    element = await fixture('<pds-list size="sm"><span/><pds-list>');
    expect(element).toMatchSnapshot();
    expect(consoleError).toHaveBeenCalledTimes(1);
    consoleError.mockRestore();
  });
});
