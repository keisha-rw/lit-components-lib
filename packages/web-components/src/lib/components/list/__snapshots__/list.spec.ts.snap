// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List unit tests does not populate the default slot with an element other than pds-list-item 1`] = `
<pds-list
  size="sm"
/>
`;
